lib/reversetunnel/watcher.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reversetunnel
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+)
+
+// RemoteClusterWatcher manages AgentPools for trusted (remote) clusters. It
+// polls the auth server for ReverseTunnel resources and spins AgentPools for
+// each one as needed.
+//
+// Note: ReverseTunnel resources on the auth server represent the desired
+// tunnels, not actually active ones.
+type RemoteClusterWatcher struct {
```

Comment:
Nit: I'm a bit hesitant about this name, just because this type is doing a lot more than what a typical `*Watcher` type does in teleport.  Would it be possible to use something a bit more descriptive (e.g. `RemoteClusterTunneler`)?

Reply:
Renamed to `RemoteClusterTunnelManager` :)